include/circt/Dialect/HW/HWTypesImpl.td

```diff
@@ -138,6 +138,31 @@ def StructTypeImpl : HWType<"Struct"> {
   }];
 }
 
+// An enum type. Declares the hw::EnumType in C++.
+def EnumTypeImpl : HWType<"Enum"> {
+  let summary = "HW Enum type";
+  let description = [{
+    Represents an enumeration of values. Enums are interpreted as integers with
+    a synthesis-defined encoding.
+    !hw.enum<field1, field2>
```

Comment:
Should enum fields have optional constant values?

Reply:
This, and above, is what i was trying to describe with

> This initial commit leaves all encoding up to the synthesis tool. If needed, the type can in the future be elaborated with optional data type (e.g. unsigned int in sv) as well as custom encodings for each enumerated value.

I do think these are important considerations, but the initial support of this commit (and the followup #3453) provides a path to synthesis for our immediate use-case so i think it'd be OK to land this first and then delve into encoding at a later point.